strawberryfields/parameters.py

```diff
@@ -19,335 +19,334 @@
+   par_regref_deps
+   par_str
+
+
+Parameter classes
 -----------------
 
-.. currentmodule:: strawberryfields.parameters.Parameter
+.. autosummary::
+   MeasuredParameter
+   FreeParameter
+
+
+Exceptions
+----------
 
 .. autosummary::
-   evaluate
+   ParameterError
 
 
 Code details
 ~~~~~~~~~~~~
 
 """
-import numbers
+# pylint: disable=too-many-ancestors,unused-argument,protected-access
+
+from collections.abc import Sequence
+import functools
+import types
 
-from unittest import mock
 import numpy as np
+import sympy
+import sympy.functions as sf
+
+
+def wrap_mathfunc(func):
+    """Applies the wrapped sympy function elementwise to numpy arrays.
+
+    Required because the sympy math functions cannot deal with arrays.
+    We implement no broadcasting; if the first argument is a numpy array, we assume
+    all the arguments are arrays of the same shape.
+    """
+    @functools.wraps(func)
+    def wrapper(*args):
+        if isinstance(args[0], np.ndarray):
+            # apply func elementwise, recursively, on the args
+            return np.array([wrapper(*k) for k in zip(*args)])
+        return func(*args)
+    return wrapper
 
-from .program_utils import (RegRef, RegRefTransform)
+# namespace of mathematical functions for manipulating Parameters
+parfuncs = types.SimpleNamespace(**{name: wrap_mathfunc(getattr(sf, name)) for name in dir(sf) if name[0] != '_'})
 
 
-try:
-    import tensorflow as tf
-    _tf_classes = (tf.Tensor, tf.Variable)
-except ImportError:
-    tf = mock.MagicMock()
-    _tf_classes = tuple()
 
+class ParameterError(RuntimeError):
+    """Exception raised when the Parameter classes encounter an illegal operation.
+
+    E.g., trying to use a measurement result before it is available.
+    """
+
+
+def is_object_array(p):
+    """Returns True iff p is an object array."""
+    return isinstance(p, np.ndarray) and p.dtype == object
 
-def _unwrap(params):
-    """Unwrap a parameter sequence.
+
+def par_evaluate(params):
+    """Evaluate a parameter sequence.
+
+    Any parameters descending from sympy.Basic are evaluated, others are returned as is.
+    Evaluation means that free and measured parameters are replaced by their numeric values.
+    Numpy arrays are evaluated elementwise.
+
+    Alternatively, evaluates a single parameter and returns its value.
 
     Args:
-      params (Sequence[Parameter]): parameters to unwrap
+        params (Sequence[Any]): parameters to evaluate
 
     Returns:
-      tuple[Number, array, Tensor, Variable]: unwrapped Parameters
+        list[Any]: evaluated parameters
+    """
+    scalar = False
+    if not isinstance(params, Sequence):
+        scalar = True
+        params = [params]
+
+    def do_evaluate(p):
+        if is_object_array(p):
+            return np.array([do_evaluate(k) for k in p])
+
+        if not par_is_symbolic(p):
+            return p
+
+        #p = p.evalf()  # TODO sympy 1.4 has some bugs which prevent using this for now, workaround below
```

Comment:
Could you add the link to the issue here, so that we know when it can be uncommented?

Reply:
I left this comment and the `if 0:` block below to show the approaches that did not work, for discussion.
I'll remove them before this PR is merged.

Reply:
It seems `sympy.lambdify` works best, it allows us to substitute anything for the atoms in the symbolic expressions.